File: movie-app/frontend/src/services/movieService.js
import axios from "axios";

const API_BASE = "http://localhost:5062/api/movies";

const normalizeMediaType = (mediaType) => (mediaType === "tv" ? "tv" : "movie");

const createMediaEndpoint = (mediaType, path) => `${API_BASE}/${normalizeMediaType(mediaType)}${path}`;

const unwrapResponseData = (data) => {
    if (typeof data !== "string") {
        return data;
    }

    try {
        return JSON.parse(data);
    } catch {
        return data;
    }
};

// =========================
// 1. POPULAR MOVIES
// =========================
export const getPopularMovies = async(page = 1) => {
    const response = await axios.get(
        `${API_BASE}/popular?page=${page}`
    );
    return unwrapResponseData(response.data);
};

export const getPopularMedia = async(mediaType, page = 1) => {
    const response = await axios.get(createMediaEndpoint(mediaType, `/popular?page=${page}`));
    return unwrapResponseData(response.data);
};

// =========================
// 2. TOP RATED
// =========================
export const getTopRatedMovies = async(page = 1) => {
    const response = await axios.get(
        `${API_BASE}/top-rated?page=${page}`
    );
    return unwrapResponseData(response.data);
};

export const getTopRatedMedia = async(mediaType, page = 1) => {
    const response = await axios.get(createMediaEndpoint(mediaType, `/top-rated?page=${page}`));
    return unwrapResponseData(response.data);
};

// =========================
// 3. UPCOMING
// =========================
export const getUpcomingMovies = async(page = 1) => {
    const response = await axios.get(
        `${API_BASE}/upcoming?page=${page}`
    );
    return unwrapResponseData(response.data);
};

export const getUpcomingMedia = async(mediaType, page = 1) => {
    const response = await axios.get(createMediaEndpoint(mediaType, `/upcoming?page=${page}`));
    return unwrapResponseData(response.data);
};

// =========================
// 4. NOW PLAYING
// =========================
export const getNowPlayingMovies = async(page = 1) => {
    const response = await axios.get(
        `${API_BASE}/now-playing?page=${page}`
    );
    return unwrapResponseData(response.data);
};

export const getNowPlayingMedia = async(mediaType, page = 1) => {
    const response = await axios.get(createMediaEndpoint(mediaType, `/now-playing?page=${page}`));
    return unwrapResponseData(response.data);
};

// =========================
// 5. GENRES
// =========================
export const getGenres = async() => {
    const response = await axios.get(
        `${API_BASE}/genres`
    );
    return unwrapResponseData(response.data);
};

export const getGenresByMedia = async(mediaType) => {
    const response = await axios.get(createMediaEndpoint(mediaType, `/genres`));
    return unwrapResponseData(response.data);
};

// =========================
// 6. MOVIES BY GENRE
// =========================
export const getMoviesByGenre = async(genreIds, page = 1) => {
    const response = await axios.get(
        `${API_BASE}/by-genre?genreIds=${genreIds}&page=${page}`
    );
    return unwrapResponseData(response.data);
};

export const getMediaByGenre = async(mediaType, genreIds, page = 1) => {
    const response = await axios.get(createMediaEndpoint(mediaType, `/by-genre?genreIds=${genreIds}&page=${page}`));
    return unwrapResponseData(response.data);
};

// =========================
// 7. SEARCH MOVIES
// =========================
export const searchMovies = async(query, page = 1) => {
    const response = await axios.get(
        `${API_BASE}/search?query=${query}&page=${page}`
    );
    return unwrapResponseData(response.data);
};

export const searchMedia = async(mediaType, query, page = 1) => {
    const response = await axios.get(createMediaEndpoint(mediaType, `/search?query=${encodeURIComponent(query)}&page=${page}`));
    return unwrapResponseData(response.data);
};

// =========================
// 8. MOVIE DETAIL
// GET: api/movies/{id}
export const getMovieDetail = async(id) => {
    const response = await axios.get(`${API_BASE}/${id}`);
    return unwrapResponseData(response.data);
};

export const getMediaDetail = async(mediaType, id) => {
    const response = await axios.get(createMediaEndpoint(mediaType, `/${id}`));
    return unwrapResponseData(response.data);
};

// =========================
// 9. MOVIE CAST (CREDITS)
// =========================
export const getMovieCredits = async(id) => {
    const response = await axios.get(`${API_BASE}/${id}/credits`);
    return unwrapResponseData(response.data);
};

export const getMediaCredits = async(mediaType, id) => {
    const response = await axios.get(createMediaEndpoint(mediaType, `/${id}/credits`));
    return unwrapResponseData(response.data);
};

// =========================
// 10. MOVIE REVIEWS
// =========================
export const getMovieReviews = async(id, page = 1) => {
    const response = await axios.get(`${API_BASE}/${id}/reviews?page=${page}`);
    return unwrapResponseData(response.data);
};

export const getMediaReviews = async(mediaType, id, page = 1) => {
    const response = await axios.get(createMediaEndpoint(mediaType, `/${id}/reviews?page=${page}`));
    return unwrapResponseData(response.data);
};

// =========================
// 11. SIMILAR MOVIES
// =========================
export const getSimilarMovies = async(id, page = 1) => {
    const response = await axios.get(`${API_BASE}/${id}/similar?page=${page}`);
    return unwrapResponseData(response.data);
};

export const getSimilarMedia = async(mediaType, id, page = 1) => {
    const response = await axios.get(createMediaEndpoint(mediaType, `/${id}/similar?page=${page}`));
    return unwrapResponseData(response.data);
};

// =========================
// 12. TRENDING MOVIES (WEEK)
// =========================
export const getTrendingMovies = async(page = 1) => {
    const response = await axios.get(`${API_BASE}/trending/week?page=${page}`);
    return unwrapResponseData(response.data);
};

export const getTrendingMedia = async(mediaType, page = 1) => {
    const response = await axios.get(createMediaEndpoint(mediaType, `/trending/week?page=${page}`));
    return unwrapResponseData(response.data);
};

// =========================
// 13. DISCOVER
// options: object of discover params (e.g., { with_genres: '18', sort_by: 'popularity.desc', page: 1 })
export const getDiscoverMedia = async(mediaType, options = {}) => {
        const params = new URLSearchParams();
        Object.entries(options || {}).forEach(([k, v]) => {
            if (v !== undefined && v !== null && v !== "") params.append(k, v);
        });
        const qs = params.toString();
        const url = createMediaEndpoint(mediaType, `/discover${qs ? `?${qs}` : ""}`);
    const response = await axios.get(url);
    return unwrapResponseData(response.data);
};

File: movie-app/backend/services/MovieService.cs
using System.Net.Http.Headers;

namespace backend.Services;

public class MovieService
{
    private readonly HttpClient _httpClient;
    private readonly IConfiguration _configuration;

    private static string NormalizeMediaType(string mediaType)
    {
        return string.Equals(mediaType, "tv", StringComparison.OrdinalIgnoreCase) ? "tv" : "movie";
    }

    public MovieService(HttpClient httpClient, IConfiguration configuration)
    {
        _httpClient = httpClient;
        _configuration = configuration;
    }

    // =========================
    // SET AUTH HEADER (reuse)
    // =========================
    private void SetAuthHeader()
    {
        var token = _configuration["TMDb:BearerToken"];

        _httpClient.DefaultRequestHeaders.Authorization =
            new AuthenticationHeaderValue("Bearer", token);

        _httpClient.DefaultRequestHeaders.Accept.Clear();
        _httpClient.DefaultRequestHeaders.Accept.Add(
            new MediaTypeWithQualityHeaderValue("application/json"));
    }

    // =========================
    // 1. POPULAR MOVIES
    // =========================
    public async Task<string> GetPopularMoviesAsync(int page = 1)
    {
        SetAuthHeader();

        var url =
            $"https://api.themoviedb.org/3/movie/popular?language=vi-VN&page={page}";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    public async Task<string> GetPopularMediaAsync(string mediaType, int page = 1)
    {
        SetAuthHeader();

        var type = NormalizeMediaType(mediaType);
        var url = $"https://api.themoviedb.org/3/{type}/popular?language=vi-VN&page={page}";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    // =========================
    // 2. TOP RATED MOVIES
    // =========================
    public async Task<string> GetTopRatedMoviesAsync(int page = 1)
    {
        SetAuthHeader();

        var url =
            $"https://api.themoviedb.org/3/movie/top_rated?language=vi-VN&page={page}";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    public async Task<string> GetTopRatedMediaAsync(string mediaType, int page = 1)
    {
        SetAuthHeader();

        var type = NormalizeMediaType(mediaType);
        var url = $"https://api.themoviedb.org/3/{type}/top_rated?language=vi-VN&page={page}";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    // =========================
    // 3. UPCOMING MOVIES
    // =========================
    public async Task<string> GetUpcomingMoviesAsync(int page = 1)
    {
        SetAuthHeader();

        var url =
            $"https://api.themoviedb.org/3/movie/upcoming?language=vi-VN&page={page}";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    public async Task<string> GetUpcomingMediaAsync(string mediaType, int page = 1)
    {
        SetAuthHeader();

        var type = NormalizeMediaType(mediaType);
        var endpoint = type == "tv" ? "airing_today" : "upcoming";
        var url = $"https://api.themoviedb.org/3/{type}/{endpoint}?language=vi-VN&page={page}";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    // =========================
    // 4. NOW PLAYING
    // =========================
    public async Task<string> GetNowPlayingMoviesAsync(int page = 1)
    {
        SetAuthHeader();

        var url =
            $"https://api.themoviedb.org/3/movie/now_playing?language=vi-VN&page={page}";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    public async Task<string> GetNowPlayingMediaAsync(string mediaType, int page = 1)
    {
        SetAuthHeader();

        var type = NormalizeMediaType(mediaType);
        var endpoint = type == "tv" ? "on_the_air" : "now_playing";
        var url = $"https://api.themoviedb.org/3/{type}/{endpoint}?language=vi-VN&page={page}";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    // =========================
    // 5. GET ALL GENRES
    // =========================
    public async Task<string> GetGenresAsync()
    {
        SetAuthHeader();

        var url =
            "https://api.themoviedb.org/3/genre/movie/list?language=vi-VN";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    public async Task<string> GetGenresAsync(string mediaType)
    {
        SetAuthHeader();

        var type = NormalizeMediaType(mediaType);
        var url = $"https://api.themoviedb.org/3/genre/{type}/list?language=vi-VN";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    // =========================
    // 6. MOVIES BY GENRE
    // =========================
    public async Task<string> GetMoviesByGenreAsync(string genreIds, int page = 1)
    {
        SetAuthHeader();

        var url =
            $"https://api.themoviedb.org/3/discover/movie" +
            $"?with_genres={genreIds}&language=vi-VN&page={page}";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    public async Task<string> GetMediaByGenreAsync(string genreIds, string mediaType, int page = 1)
    {
        SetAuthHeader();

        var type = NormalizeMediaType(mediaType);
        var url = $"https://api.themoviedb.org/3/discover/{type}?with_genres={genreIds}&language=vi-VN&page={page}";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    // =========================
    // 7. SEARCH MOVIE
    // =========================
    public async Task<string> SearchMoviesAsync(string keyword, int page = 1)
    {
        SetAuthHeader();

        var url =
            $"https://api.themoviedb.org/3/search/movie" +
            $"?query={Uri.EscapeDataString(keyword)}&language=vi-VN&page={page}";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    public async Task<string> SearchMediaAsync(string keyword, string mediaType, int page = 1)
    {
        SetAuthHeader();

        var type = NormalizeMediaType(mediaType);
        var url = $"https://api.themoviedb.org/3/search/{type}?query={Uri.EscapeDataString(keyword)}&language=vi-VN&page={page}";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    // =========================
    // 8. MOVIE DETAIL
    // =========================
    public async Task<string> GetMovieDetailAsync(int movieId)
    {
        SetAuthHeader();

        var url =
            $"https://api.themoviedb.org/3/movie/{movieId}?language=vi-VN";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    public async Task<string> GetMediaDetailAsync(int id, string mediaType)
    {
        SetAuthHeader();

        var type = NormalizeMediaType(mediaType);
        var url = $"https://api.themoviedb.org/3/{type}/{id}?language=vi-VN";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }


    // =========================
    // 9. MOVIE CREDITS (CAST + CREW)
    // =========================
    public async Task<string> GetMovieCreditsAsync(int id)
    {
        SetAuthHeader();

        var url = $"https://api.themoviedb.org/3/movie/{id}/credits";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    public async Task<string> GetMediaCreditsAsync(int id, string mediaType)
    {
        SetAuthHeader();

        var type = NormalizeMediaType(mediaType);
        var url = $"https://api.themoviedb.org/3/{type}/{id}/credits";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    // =========================
    // 10. MOVIE REVIEWS
    // =========================
    public async Task<string> GetMovieReviewsAsync(int id, int page = 1)
    {
        SetAuthHeader();

        var url =
            $"https://api.themoviedb.org/3/movie/{id}/reviews?language=vi-VN&page={page}";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    public async Task<string> GetMediaReviewsAsync(int id, string mediaType, int page = 1)
    {
        SetAuthHeader();

        var type = NormalizeMediaType(mediaType);
        var url = $"https://api.themoviedb.org/3/{type}/{id}/reviews?language=vi-VN&page={page}";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    // =========================
    // 11. SIMILAR MOVIES
    // =========================
    public async Task<string> GetSimilarMoviesAsync(int id, int page = 1)
    {
        SetAuthHeader();

        var url =
            $"https://api.themoviedb.org/3/movie/{id}/similar?language=vi-VN&page={page}";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    public async Task<string> GetSimilarMediaAsync(int id, string mediaType, int page = 1)
    {
        SetAuthHeader();

        var type = NormalizeMediaType(mediaType);
        var url = $"https://api.themoviedb.org/3/{type}/{id}/similar?language=vi-VN&page={page}";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    // =========================
    // 12. TRENDING MOVIES (WEEK)
    // =========================
    public async Task<string> GetTrendingMoviesAsync(int page = 1)
    {
        SetAuthHeader();

        var url =
            $"https://api.themoviedb.org/3/trending/movie/week?language=vi-VN&page={page}";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    public async Task<string> GetTrendingMediaAsync(string mediaType, int page = 1)
    {
        SetAuthHeader();

        var type = NormalizeMediaType(mediaType);
        var url = $"https://api.themoviedb.org/3/trending/{type}/week?language=vi-VN&page={page}";

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }

    // =========================
    // 13. DISCOVER (GENERAL)
    // Accepts arbitrary discover query parameters forwarded from client
    // Example: /api/movies/tv/discover?with_genres=18&sort_by=popularity.desc&page=1
    // =========================
    public async Task<string> DiscoverMediaAsync(string mediaType, string queryString)
    {
        SetAuthHeader();

        var type = NormalizeMediaType(mediaType);

        // queryString includes leading '?' when forwarded from HttpContext.Request.QueryString
        var qs = string.IsNullOrWhiteSpace(queryString) ? string.Empty : queryString.TrimStart('?');

        var url = $"https://api.themoviedb.org/3/discover/{type}?language=vi-VN" + (string.IsNullOrEmpty(qs) ? string.Empty : "&" + qs);

        var response = await _httpClient.GetAsync(url);
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }
}